jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/SingleSourceAllocation.kt

```diff
@@ -0,0 +1,329 @@
+    constraints: VideoConstraints,
+    onStage: Boolean
+): Layers {
+
+    var activeLayers = layers.filter { it.bitrate > 0 }
+    // No active layers usually happens when the source has just been signaled and we haven't received
+    // any packets yet. Add the layers here, so one gets selected and we can start forwarding sooner.
+    if (activeLayers.isEmpty()) activeLayers = layers
+
+    // We select all layers that satisfy the constraints.
+    var selectedLayers =
+        if (constraints.maxHeight < 0) {
+            activeLayers
+        } else {
+            activeLayers.filter { it.layer.height <= constraints.maxHeight }
+        }
+    // If no layers satisfy the constraints, we use the layers with the lowest resolution.
+    if (selectedLayers.isEmpty()) {
+        val minHeight = activeLayers.map { it.layer.height }.minOrNull() ?: return noLayers
+        selectedLayers = activeLayers.filter { it.layer.height == minHeight }
+    }
+
+    val oversendIdx = if (onStage && BitrateControllerConfig.allowOversendOnStage()) {
+        val maxHeight = selectedLayers.map { it.layer.height }.maxOrNull() ?: return noLayers
+        selectedLayers.firstIndexWhich { it.layer.height == maxHeight }
+    } else {
+        -1
+    }
+    return Layers(selectedLayers, selectedLayers.size - 1, oversendIdx)
+}
+
+private fun <T> List<T>.firstIndexWhich(predicate: (T) -> Boolean): Int {
+    forEachIndexed { index, item ->
+        if (predicate(item)) return index
+    }
+    return -1
+}
+
+private fun selectLayersForCamera(
+    layers: List<LayerSnapshot>,
+    constraints: VideoConstraints,
+): Layers {
+
+    val minHeight = layers.map { it.layer.height }.minOrNull() ?: return noLayers
+    val noActiveLayers = layers.none { (_, bitrate) -> bitrate > 0 }
+    val (preferredHeight, preferredFps) = getPreferred(constraints)
+
+    val ratesList: MutableList<LayerSnapshot> = ArrayList()
+    // Initialize the list of layers to be considered. These are the layers that satisfy the constraints, with
+    // a couple of exceptions (see comments below).
+    for (layerSnapshot in layers) {
+        val layer = layerSnapshot.layer
+        val lessThanPreferredHeight = layer.height < preferredHeight
+        val lessThanOrEqualMaxHeight = layer.height <= constraints.maxHeight
+        // If frame rate is unknown, consider it to be sufficient.
+        val atLeastPreferredFps = layer.frameRate < 0 || layer.frameRate >= preferredFps
+        if (lessThanPreferredHeight ||
+            (lessThanOrEqualMaxHeight && atLeastPreferredFps) ||
+            layer.height == minHeight
+        ) {
+
+            // No active layers usually happens when the source has just been signaled and we haven't received
+            // any packets yet. Add the layers here, so one gets selected and we can start forwarding sooner.
+            if (noActiveLayers || layerSnapshot.bitrate > 0) {
+                ratesList.add(layerSnapshot)
+            }
+        }
+    }
```

Comment:
The old code had explicit logic that guaranteed that `ratesList` would have at least one element (assuming `layers` does).  Does this new logic guarantee that too?  I could believe it, but I'd like to be certain.

Reply:
I think with the new code there is one case in which this is not true, and it is intentional. This is the case of the receiver signaling maxHeight=0. The old code adds a layer, the new code doesn't. This commit makes the change, and introduces tests for it:
https://github.com/jitsi/jitsi-videobridge/pull/1687/commits/636694da6c87156fe68112ebdba5f711c8d1c676

There is one more related change: when maxHeight!=0, and all available layers have height>maxHeight. The old code considered a single layer, the new code considers all layers with the lowest res. For example, consider a non-simulcast sender sending 720p and a receiver signaling maxHeight=360 (e.g. tile view). The old code would consider only the low fps layer, the new code considers them all.